File: CSG.py
class CSG(object):
    def __init__(self):
        self.rotation = np.eye(2)
        self.translation = np.zeros(2)
    def __call__(self, pos):
        if not hasattr(self, '_map'):
            raise ValueError('CSG solid does not implement map function')
        return self._map(self._transform(pos))
    def _transform(self, pos):
        return self.rotation.dot(pos - self.translation)
    def set_rotation(self, angle):
        c = np.cos(-angle)
        s = np.sin(-angle)
        self.rotation = np.array([[c, -s], [s, c]])
    def set_translation(self, offset):
        self.translation = np.array(offset)

class Union(CSG):
    def __init__(self, *shapes):
        super(Union, self).__init__()
        self.shapes = shapes
    def _map(self, pos):
        return min([shape(pos) for shape in self.shapes])

class Intersection(CSG):
    def __init__(self, *shapes):
        super(Intersection, self).__init__()
        self.shapes = shapes
    def _map(self, pos):
        return max([shape(pos) for shape in self.shapes])

class Negation(CSG):
    def __init__(self, shape):
        super(Negation, self).__init__()
        self.shape = shape
    def _map(self, pos):
        return -self.shape(pos)

class Circle(CSG):
    def __init__(self, radius):
        super(Circle, self).__init__()
        self.radius = radius
    def _map(self, pos):
        return np.sqrt(pos[0]*pos[0]+pos[1]*pos[1]) - self.radius

class Square(CSG):
    def __init__(self, radius):
        super(Square, self).__init__()
        self.radius = radius
    def _map(self, pos):
        inside = np.max(np.abs(pos)) - self.radius
        disp = np.clip(np.abs(pos)-self.radius, 0, None)
        return min(inside, 0) + np.sqrt(disp.dot(disp))

File: CSG.jl
include("typedefs.jl")
struct Transform{T, C<:CSG{T}} <: CSG{T}
    child :: C
    x :: T
    y :: T
    matrix :: Matrix{T}
    Transform{T}(child::C, x::Real, y::Real, m::Matrix) where {T, C<:CSG{T}} = new{T,C}(child, x, y, m^-1)
end
function rotmatrix(r::Real)
    c = cos(r)
    s = sin(r)
    return [c s; -s c]
end
Transform(child::CSG{T}, x::Real, y::Real, r::Real=0.0) where {T} = Transform{T}(child, x, y, rotmatrix(r))


struct Circle{T} <: CSG{T}
    radius :: T
end

struct Square{T} <: CSG{T}
    radius :: T
end

#composite CSG types
struct Intersect{T} <: CSG{T}
    children :: Vector{CSG{T}}
end
struct SoftIntersect{T} <: CSG{T}
    children :: Vector{CSG{T}}
    smoothness :: T
end
struct Union{T} <: CSG{T}
    children :: Vector{CSG{T}}
end
struct Union2{T, C<:CSG{T}, D<:CSG{T}} <: CSG{T}
    left :: C
    right :: D
end
struct SoftUnion{T} <: CSG{T}
    children :: Vector{CSG{T}}
    smoothness :: T
end
struct Negate{T, C<:CSG{T}} <: CSG{T}
    child :: C
end
struct UniformSolid{T} <: CSG{T}
    filled :: Bool
end
struct Displace{T, C<:CSG{T}, F<:Field{T}} <: CSG{T}
    child :: C
    selection :: F
    factor :: T
end


import Base.getindex
function getindex(trans::Transform{T}, posx::T, posy::T) where {T}
    posx, posy = (posx-trans.x, posy-trans.y)
    newposx, newposy = trans.matrix[1,1] * posx + trans.matrix[1,2] * posy, trans.matrix[2,1] * posx + trans.matrix[2,2] * posy
    return trans.child[newposx, newposy]
end
#signed distance functions
getindex(csg::Circle{T}, posx::T, posy::T) where {T} = sqrt(posx*posx+posy*posy) - csg.radius
function getindex(csg::Square{T}, pos::Vararg{T, 2}) where {T}
    inside = max(abs.(pos)...) - csg.radius
    disp = max.(abs.(pos) .- csg.radius, 0)
    return min(inside, 0) + sqrt(disp[1]*disp[1]+disp[2]*disp[2])
end
getindex(csg::Intersect{T}, pos::Vararg{T, 2}) where {T} = maximum(child[pos...] for child in csg.children)
getindex(csg::SoftIntersect{T}, pos::Vararg{T, 2}) where {T} = csg.smoothness*log(sum(exp(child[pos...]/csg.smoothness) for child in csg.children))
getindex(csg::Union{T}, pos::Vararg{T, 2}) where {T} = minimum(child[pos...] for child in csg.children)
getindex(csg::Union2{T}, pos::Vararg{T,2}) where {T} = min(csg.left[pos...], csg.right[pos...])
getindex(csg::SoftUnion{T}, pos::Vararg{T, 2}) where {T} = -csg.smoothness*log(sum(exp(-child[pos...]/csg.smoothness) for child in csg.children))
getindex(csg::Negate{T}, pos::Vararg{T, 2}) where {T} = -csg.child[pos...]
getindex(csg::UniformSolid{T}, pos::Vararg{T, 2}) where {T} = csg.filled ? typemin(T) : typemax(T)
getindex(csg::Displace{T}, pos::Vararg{T, 2}) where {T} = csg.child[pos...] - csg.selection[pos...] * csg.factor
